File: src/sync.ts
/**
 * S3 vault sync — push/pull SQLite vault + vector store to/from S3.
 *
 * Uses AWS CLI (no SDK dependency). Content-addressable storage
 * prevents content duplication. Last-write-wins for metadata.
 *
 * Usage:
 *   MNEMON_S3_BUCKET=my-bucket bun run src/sync.ts push
 *   MNEMON_S3_BUCKET=my-bucket bun run src/sync.ts pull
 */

import { existsSync, statSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";

const S3_BUCKET = process.env.MNEMON_S3_BUCKET ?? "";
const S3_PREFIX = process.env.MNEMON_S3_PREFIX ?? "mnemon/vaults";
const VAULT_NAME = process.env.MNEMON_VAULT_NAME ?? "default";

function vaultDir(): string {
  return join(homedir(), ".mnemon");
}

function vaultFiles(): { sqlite: string; vec: string } {
  const dir = vaultDir();
  return {
    sqlite: join(dir, `${VAULT_NAME}.sqlite`),
    vec: join(dir, `${VAULT_NAME}.vec`),
  };
}

function s3Path(filename: string): string {
  return `s3://${S3_BUCKET}/${S3_PREFIX}/${filename}`;
}

async function runCmd(cmd: string): Promise<{ ok: boolean; output: string }> {
  const proc = Bun.spawn(["bash", "-c", cmd], {
    stdout: "pipe",
    stderr: "pipe",
  });

  const stdout = await new Response(proc.stdout).text();
  const stderr = await new Response(proc.stderr).text();
  const exitCode = await proc.exited;

  return {
    ok: exitCode === 0,
    output: exitCode === 0 ? stdout.trim() : stderr.trim(),
  };
}

/**
 * Push local vault to S3.
 */
export async function push(): Promise<{ pushed: string[]; errors: string[] }> {
  if (!S3_BUCKET) {
    return { pushed: [], errors: ["MNEMON_S3_BUCKET not set"] };
  }

  const files = vaultFiles();
  const pushed: string[] = [];
  const errors: string[] = [];

  for (const [label, localPath] of Object.entries(files)) {
    if (!existsSync(localPath)) continue;

    const s3Target = s3Path(`${VAULT_NAME}.${label === "sqlite" ? "sqlite" : "vec"}`);
    const result = await runCmd(`aws s3 cp "${localPath}" "${s3Target}" --only-show-errors`);

    if (result.ok) {
      const size = statSync(localPath).size;
      pushed.push(`${label}: ${(size / 1024).toFixed(1)}KB → ${s3Target}`);
    } else {
      errors.push(`${label}: ${result.output}`);
    }
  }

  return { pushed, errors };
}

/**
 * Pull vault from S3 to local.
 */
export async function pull(): Promise<{ pulled: string[]; errors: string[] }> {
  if (!S3_BUCKET) {
    return { pulled: [], errors: ["MNEMON_S3_BUCKET not set"] };
  }

  const files = vaultFiles();
  const pulled: string[] = [];
  const errors: string[] = [];

  for (const [label, localPath] of Object.entries(files)) {
    const s3Source = s3Path(`${VAULT_NAME}.${label === "sqlite" ? "sqlite" : "vec"}`);

    // Check if file exists on S3
    const exists = await runCmd(`aws s3 ls "${s3Source}" 2>/dev/null`);
    if (!exists.ok || !exists.output) continue;

    const result = await runCmd(`aws s3 cp "${s3Source}" "${localPath}" --only-show-errors`);

    if (result.ok) {
      const size = existsSync(localPath) ? statSync(localPath).size : 0;
      pulled.push(`${label}: ${s3Source} → ${(size / 1024).toFixed(1)}KB`);
    } else {
      errors.push(`${label}: ${result.output}`);
    }
  }

  return { pulled, errors };
}

// ── CLI ─────────────────────────────────────────────────────────────────────

if (import.meta.main) {
  const action = process.argv[2];

  if (action === "push") {
    const result = await push();
    if (result.pushed.length > 0) {
      console.log("Pushed:");
      result.pushed.forEach((p) => console.log(`  ${p}`));
    }
    if (result.errors.length > 0) {
      console.error("Errors:");
      result.errors.forEach((e) => console.error(`  ${e}`));
      process.exit(1);
    }
  } else if (action === "pull") {
    const result = await pull();
    if (result.pulled.length > 0) {
      console.log("Pulled:");
      result.pulled.forEach((p) => console.log(`  ${p}`));
    }
    if (result.errors.length > 0) {
      console.error("Errors:");
      result.errors.forEach((e) => console.error(`  ${e}`));
      process.exit(1);
    }
  } else {
    console.log("Usage: bun run src/sync.ts <push|pull>");
    console.log("\nEnv vars:");
    console.log("  MNEMON_S3_BUCKET  — S3 bucket name (required)");
    console.log("  MNEMON_S3_PREFIX  — S3 key prefix (default: mnemon/vaults)");
    console.log("  MNEMON_VAULT_NAME — vault name (default: default)");
  }
}

File: src/mnemon/sync.py
"""S3 vault sync — push/pull SQLite vault + vector store to/from S3.

Uses AWS CLI (no SDK dependency). Content-addressable storage
prevents content duplication. Last-write-wins for metadata.

Usage:
    MNEMON_S3_BUCKET=my-bucket mnemon sync push
    MNEMON_S3_BUCKET=my-bucket mnemon sync pull

Env vars:
    MNEMON_S3_BUCKET    S3 bucket name (required)
    MNEMON_S3_PREFIX    S3 key prefix (default: mnemon/vaults)
    MNEMON_VAULT_NAME   vault name (default: default)
"""

from __future__ import annotations

import os
import subprocess
from pathlib import Path

from .config import vault_dir

S3_PREFIX_DEFAULT = "mnemon/vaults"
VAULT_NAME_DEFAULT = "default"


def _s3_bucket() -> str:
    return os.environ.get("MNEMON_S3_BUCKET", "")


def _s3_prefix() -> str:
    return os.environ.get("MNEMON_S3_PREFIX", S3_PREFIX_DEFAULT)


def _vault_name() -> str:
    return os.environ.get("MNEMON_VAULT_NAME", VAULT_NAME_DEFAULT)


def _vault_files() -> dict[str, Path]:
    """Return local vault file paths."""
    vdir = vault_dir()
    name = _vault_name()
    return {
        "sqlite": vdir / f"{name}.sqlite",
        "vec": vdir / f"{name}.vec.npz",
    }


def _s3_path(filename: str) -> str:
    return f"s3://{_s3_bucket()}/{_s3_prefix()}/{filename}"


def _run_cmd(cmd: str) -> tuple[bool, str]:
    """Run a shell command. Returns (success, output)."""
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    if result.returncode == 0:
        return True, result.stdout.strip()
    return False, result.stderr.strip()


def _checkpoint_wal(sqlite_path: Path) -> str | None:
    """Force a WAL checkpoint so the main sqlite file contains all
    committed writes before ``aws s3 cp`` reads it.

    Short-lived CLI processes (e.g. ``mnemon save``) auto-checkpoint
    on connection close, so the WAL file disappears and the main file
    contains all data. Long-lived processes (e.g. ``mnemon
    serve-remote`` on Fly) hold an open SQLite connection — new
    commits accumulate in the WAL and SQLite only auto-checkpoints
    once the WAL grows past 1000 pages (default). Without an explicit
    checkpoint here, ``mnemon sync push`` running against a
    long-running server uploads a stale main file missing the latest
    writes, and the downstream ``sync pull`` silently restores
    the stale state.

    Surfaced 2026-05-21 during the 0.6.0 Layer-3 test: Step 4 added a
    doc via remote (committed to Fly serve-remote's WAL but never
    flushed to main); downgrade's Fly→S3 dump (added in 0.6.0 too)
    then uploaded the stale main and lost the doc on local restore.

    TRUNCATE mode: most thorough, blocks briefly if other readers
    hold locks. Acceptable for the (push, sync) flow — we don't
    expect concurrent traffic during a deliberate sync push. Failures
    are returned as a string so the caller can log without raising
    (sync push remains best-effort per error).
    """
    import sqlite3
    try:
        conn = sqlite3.connect(str(sqlite_path), timeout=10.0)
        try:
            conn.execute("PRAGMA wal_checkpoint(TRUNCATE);")
            conn.commit()
        finally:
            conn.close()
    except sqlite3.Error as exc:
        return f"sqlite checkpoint failed: {exc}"
    return None


def push() -> dict[str, list[str]]:
    """Push local vault to S3.

    Returns {"pushed": [...], "errors": [...]}.
    """
    bucket = _s3_bucket()
    if not bucket:
        return {"pushed": [], "errors": ["MNEMON_S3_BUCKET not set"]}

    files = _vault_files()
    pushed: list[str] = []
    errors: list[str] = []

    for label, local_path in files.items():
        if not local_path.exists():
            continue

        # Force WAL → main flush before reading the sqlite file as bytes.
        # See _checkpoint_wal docstring for the rationale.
        if label == "sqlite":
            ckpt_err = _checkpoint_wal(local_path)
            if ckpt_err is not None:
                errors.append(ckpt_err)
                continue

        ext = "sqlite" if label == "sqlite" else "vec.npz"
        s3_target = _s3_path(f"{_vault_name()}.{ext}")
        ok, output = _run_cmd(f'aws s3 cp "{local_path}" "{s3_target}" --only-show-errors')

        if ok:
            size_kb = local_path.stat().st_size / 1024
            pushed.append(f"{label}: {size_kb:.1f}KB → {s3_target}")
        else:
            errors.append(f"{label}: {output}")

    return {"pushed": pushed, "errors": errors}


def pull() -> dict[str, list[str]]:
    """Pull vault from S3 to local.

    Returns {"pulled": [...], "errors": [...]}.
    """
    bucket = _s3_bucket()
    if not bucket:
        return {"pulled": [], "errors": ["MNEMON_S3_BUCKET not set"]}

    files = _vault_files()
    pulled: list[str] = []
    errors: list[str] = []

    for label, local_path in files.items():
        ext = "sqlite" if label == "sqlite" else "vec.npz"
        s3_source = _s3_path(f"{_vault_name()}.{ext}")

        # Check if file exists on S3
        ok, output = _run_cmd(f'aws s3 ls "{s3_source}" 2>/dev/null')
        if not ok or not output:
            continue

        # Ensure parent dir exists
        local_path.parent.mkdir(parents=True, exist_ok=True)

        ok, output = _run_cmd(f'aws s3 cp "{s3_source}" "{local_path}" --only-show-errors')

        if ok:
            size_kb = local_path.stat().st_size / 1024 if local_path.exists() else 0
            pulled.append(f"{label}: {s3_source} → {size_kb:.1f}KB")
        else:
            errors.append(f"{label}: {output}")

    return {"pulled": pulled, "errors": errors}
